docs/madr/decisions/057-rework-mesh-routes-matching.md

```diff
@@ -0,0 +1,379 @@
+# Rework of Mesh*Routes matching
+
+* Status: accepted
+
+Technical story: https://github.com/kumahq/kuma/issues/10247
+
+## Context and Problem Statement
+
+At the moment when targeting Mesh*Route you need to put it in topLevel targetRef. 
+We want to iterate over policy design to fully utilize the fact that we are referencing real objects.
+The current implementation has made things a little tricky because policy matching is convoluted where the top level 
+MeshHTTPRoute gets converted to a fake `to`.
+
+## Considered options
+
+- add possibility to put Mesh*Routes in `spec.to[].targetRef`
+- leave Mesh*Routes in top level targetRef as it is now
+
+## Decision Outcome
+
+- add possibility to put Mesh*Routes in `spec.to[].targetRef`
+
+### Positive consequences
+
+- improved policy validation, we can enforce applying only http config on MeshHttpRoute
+- simplifying topLevel targetRef as it will now match only whole proxies
+- no need for specifying topLevel targetRef when targeting `Mesh`
+- easier policy matching code
+
+## Design
+
+The design is grounded in multiple recent MADRs around improvements to policies: 
+
+- [MeshService policy matching](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/046-meshservice-policy-matching.md)
+- [Policy on Namespace](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/052-policy-on-namespace.md) - which introduced consumer/producer policies
+- [Policy matching with real resources](https://github.com/kumahq/kuma/pull/10726/files)
+
+### API
+
+New API for policy that selects Mesh*Route
+
+```yaml
+type: MeshTimeout
+name: timeouts-on-route
+mesh: default
+spec:
+  targetRef:
+    kind: Mesh | MeshSubset # either Mesh, MeshSubset or MeshGateway
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute | MeshTCPRoute # either MeshHTTPRoute or MeshTCPRoute
+        name: route-to-backend
+      default:
+        http:
+          requestTimeout: 5s
+          streamIdleTimeout: 1h
+```
+
+specifying `Mesh` in topLevel targetRef will be optional (this probably will be most used case). Minimal policy, targeting
+`MeshHTTPRoute` will look like this:
+
+```yaml
+type: MeshTimeout
+name: timeouts-on-route
+mesh: default
+spec:
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute
+        name: route-to-backend
+      default:
+        http:
+          requestTimeout: 5s
+          streamIdleTimeout: 1h
+```
+
+To introduce this change we need to deprecate specifying Mesh*Route in topLevel targetRef.
+
+Should we allow selecting MeshGateway in topLevel targetRef? probably this is obsolete as Route will select gateway, and we
```

Comment:
I think it is simple, you create route on MeshGateway:
```yaml
kind: MeshHTTPRoute
metadata:
  name: demo-app-route
  namespace: kuma-system
  labels:
    kuma.io/origin: zone
    kuma.io/mesh: default
spec:
  targetRef:
    kind: MeshGateway
    name: demo-app
  to:
    - targetRef:
        kind: Mesh
      rules:
        - matches:
            - path:
                type: PathPrefix
                value: "/"
          default:
            backendRefs:
              - kind: MeshService
                name: demo-app_kuma-demo_svc_5000
```

This route is applied on MeshGateway and then you create MeshTimeout policy like: 

```yaml
kind: MeshTimeout
metadata:
  name: demo-app-timeout
  namespace: kuma-system
  labels:
    kuma.io/origin: zone
    kuma.io/mesh: default
spec:
  to:
    - targetRef:
        kind: MeshHTTPRoute
        name: demo-app-route
      default:
        requestTimeout: 5s
```

This should be applied on Gateway on route since you target Mesh in topLevel targetRef on MeshTimeout

Reply:
Of course we need MeshGateway in the top level. There is just as much a need for MeshGateway top level as there is for MeshSubset.

1. How do I apply configuration for all routes?
2. Are we even sure that we can say whether _all_ Envoy configuration belongs to a route or not?

Reply:
This point is not about removing MeshGateway from topLevel targetRef entirely, but to not use it with Mesh*Route in to section. 

Ad 1. in this case, you will have topLevel targetRef it to you will most likely have `Mesh` or `MeshService`
Ad 2. Only part of Envoy config belongs to a route, at the moment it is requestTimeout and requestHeaderTimeout

Reply:
OK, understood. Well as far as I understand the whole point of having `spec.targetRef` is to select which proxies the policy applies to. From that point of view I don't see why you'd want to forbid it. But maybe we can evaluate in general which top level targetRefs need to be available.